File: frontend/src/api/inventory.ts
import { apiClient } from './client';

export interface InventoryItem {
    id: string;
    name: string;
    unit: string;
    current_stock: number;
    min_stock_level: number;
    unit_cost: number;
    created_at: string;
    updated_at: string;
}

export interface StockMovement {
    id: string;
    item_id: string;
    movement_type: 'purchase' | 'usage' | 'adjustment';
    quantity: number;
    reference_id?: string;
    notes?: string;
    created_by: string;
    created_at: string;
}

export const inventoryApi = {
    getAll: async (lowStockOnly = false) => {
        const response = await apiClient.get<InventoryItem[]>('/inventory', {
            params: { low_stock_only: lowStockOnly }
        });
        return response.data;
    },

    getLowStock: async () => {
        const response = await apiClient.get<InventoryItem[]>('/inventory/low-stock');
        return response.data;
    },

    create: async (data: Partial<InventoryItem>) => {
        const response = await apiClient.post<InventoryItem>('/inventory', data);
        return response.data;
    },

    update: async (id: string, data: Partial<InventoryItem>) => {
        const response = await apiClient.patch<InventoryItem>(`/inventory/${id}`, data);
        return response.data;
    },

    adjustStock: async (data: { item_id: string; movement_type: string; quantity: number; notes?: string }) => {
        const response = await apiClient.post<StockMovement>('/inventory/movements', data);
        return response.data;
    },

    getMovements: async (itemId: string) => {
        const response = await apiClient.get<StockMovement[]>(`/inventory/movements/${itemId}`);
        return response.data;
    }
};

File: backend/app/models/inventory.py
from sqlalchemy import Column, String, DateTime, Numeric, Enum, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid
import enum
from ..database import Base


class MovementType(str, enum.Enum):
    PURCHASE = "purchase"
    USAGE = "usage"
    ADJUSTMENT = "adjustment"


class InventoryItem(Base):
    __tablename__ = "inventory_items"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False, index=True)
    unit = Column(String(50), nullable=False)
    current_stock = Column(Numeric(10, 2), nullable=False, default=0)
    min_stock_level = Column(Numeric(10, 2), nullable=False, default=0)
    unit_cost = Column(Numeric(10, 2), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    
    # Future Hooks
    tenant_id = Column(String(50), nullable=True, index=True)
    
    # Relationships
    movements = relationship("StockMovement", back_populates="item")
    services = relationship("Service", back_populates="stock_item")


class StockMovement(Base):
    __tablename__ = "stock_movements"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    item_id = Column(UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False)
    movement_type = Column(Enum(MovementType), nullable=False)
    quantity = Column(Numeric(10, 2), nullable=False)
    reference_id = Column(String(255), nullable=True)
    notes = Column(String(500))
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False, index=True)
    
    # Future Hooks
    tenant_id = Column(String(50), nullable=True, index=True)
    
    # Relationships
    item = relationship("InventoryItem", back_populates="movements")
    user = relationship("User")
